cms/djangoapps/contentstore/tasks.py

```diff
@@ -1117,101 +1120,155 @@ def validate_user():
+        """
+        Returns the status of a url request
+        Returns: {block_id1, url1, status}
+        """
+        block_id, url = url_data
+        result = {'block_id': block_id, 'url': url}
+        standardized_url = convert_to_standard_url(url, course_key)
+        try:
+            async with session.get(standardized_url, timeout=5) as response:
+                result.update({'status': response.status})
+        except Exception as e:
+            result.update({'status': None})
+            LOGGER.debug(f'[Link Check] Request error when validating {url}: {str(e)}')
+        return result
+
+    async def validate_urls_access_in_batches(url_list, course_key, batch_size=100):
+        """
+        Returns the statuses of a list of url requests.
+        Returns: [ {block_id1, url1, status}, {block_id2, url2, status}, ... ]
+        """
+        responses = []
+        url_count = len(url_list)
+
+        for i in range(0, url_count, batch_size):
+            batch = url_list[i:i + batch_size]
+            async with aiohttp.ClientSession() as session:
+                tasks = [validate_url_access(session, url_data, course_key) for url_data in batch]
+                batch_results = await asyncio.gather(*tasks)
+                responses.extend(batch_results)
+                LOGGER.debug(f'[Link Check] request batch {i // batch_size+1} of {url_count // batch_size + 1}')
+        
+        return responses
+
+    def filter_by_status(results):
+        """
+        Filter results by status.
+            200: OK. No need to do more
+            403: Forbidden. Record as locked link.
+            None: Error. Retry up to 3 times.
+            Other: Failure. Record as broken link.
+        Returns:
+            filtered_results: [ [block_id1, url1, is_locked], ... ]
+            retry_list: [ [block_id1, url1], ... ]
+        """
+        filtered_results = []
+        retry_list = []
+        for result in results:
+            if result['status'] == None:
+                retry_list.append([result['block_id'], result['url']])
+            elif result['status'] == 200:
+                continue
+            elif result['status'] == 403 and is_studio_url(result['url']):
+                filtered_results.append([result['block_id'], result['url'], True])
+            else:
+                filtered_results.append([result['block_id'], result['url'], False])
+        
+        return filtered_results, retry_list
 
     user = validate_user()
 
     self.status.set_state('Scanning')
-    courselike_key = CourseKey.from_string(course_key_string)
-    data = scan_course(courselike_key)
+    course_key = CourseKey.from_string(course_key_string)
+    url_list = scan_course_for_links(course_key)
+    validated_url_list = asyncio.run(validate_urls_access_in_batches(url_list, course_key, batch_size=100))
+    broken_or_locked_urls, retry_list = filter_by_status(validated_url_list)
+
+    # Retry urls that failed due to connection error
```

Comment:
nit: Generally it's really nice that you decoupled everything into functions. Everything is extremely well readable until here, where it is still good but a bit less clear structured. A small improvement could be to put the two chunks of code that are following into functions also. The for loop is a nice thing to put into a named function so people can see what it does without needing to read the code, and in the try/except statement, everything inside of try can be a function that's called with one line in the try. That's a pattern you often see where any try/except statement will have the try block just be one function call and the except block stay as it's here. What do you think?

The code already looks really good so this is a pretty minor thing. If you prefer it like this that's okay as well.

Reply:
That makes sense to do! I'll make that change